### PR TITLE
Use strategy pattern for conditional online time series indexing

### LINKED ISSUE #70: Extract Cassandra time series indexing into an online and an "offline" strategy
All time series need to be indexed in order for them to be exposed. We currently do not do this anywhere however the functionality is partially implemented in `CassandraStore#writeTimeseriesIdIndex`. This is not called on `#addPoint` by default due to all the extra database calls it would cause. We should make this behavior configurable so that if one wish to index the time series on every call to `#addPoint` one can, but there should also be the possibility of doing this in an "offline" fashion.

This could be implemented through an "online" and an "offline" indexing strategy variable which picks between to different implementations for either case. See `se.tre.freki.labels.IdLookupStrategy` for an example of a similar implementation.

### PR DESCRIPTION
**Pull request by luuse**
Fixes #70

### CODE REVIEW

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/CassandraStore.java [hunk 1]
```diff
@@ -106,6 +104,8 @@
   private PreparedStatement getNameStatement;
   private PreparedStatement getIdStatement;
 
+  private final IndexStrategy onlineIndexingStrategy;
```
Comment (luuse) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/CassandraStore.java]: I don't like this name but it is the best I could think of. Any suggestions?

File: cassandra/src/test/java/se/tre/freki/storage/cassandra/CassandraStoreTest.java [hunk 2]
```diff
@@ -66,13 +70,17 @@ public void constructorNullCluster() throws IOException {
...
   }
 
   @Test(expected = NullPointerException.class)
   public void constructorNullSession() throws IOException {
     final Cluster cluster = storeDescriptor.createCluster(config);
-    new CassandraStore(cluster, null, Clock.systemDefaultZone());
+    final Clock clock = Clock.systemDefaultZone();
+    final IndexStrategy indexingStrategy = new NoOpIndexingStrategy();
+    new CassandraStore(cluster, null, clock, indexingStrategy);
   }
```
Comment (jzeeck) [inline cassandra/src/test/java/se/tre/freki/storage/cassandra/CassandraStoreTest.java]: Test for when indexingStrategy is null is missing.

File: cassandra/src/main/resources/reference.conf [hunk 3]
```diff
@@ -5,3 +5,6 @@ freki.storage.cassandra.protocolVersion = 2
 # Default keyspace to be used in cassandra
 freki.storage.cassandra.keyspace = "freki"
 
+# Whether time series IDs should be indexed every time a new point is added (true) or if they should
+# be indexed in a more manual fashion (false).
+freki.storage.cassandra.online_index = false
```
Comment (jzeeck) [inline cassandra/src/main/resources/reference.conf]: Maybe change this to auto_index?

Right now I do not see the connection to "online"

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/CassandraStore.java (hunk 1 above)
Comment (jzeeck) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/CassandraStore.java]: Why not just call it indexingStrategy?

File: cassandra/src/test/java/se/tre/freki/storage/cassandra/IndexingStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    final LabelId metric = fromLong(1L);
+
+    final ImmutableList<LabelId> tags = ImmutableList.<LabelId>of(
+        fromLong(2L));
+
+    final ByteBuffer timeSeriesId = TimeSeriesIds.timeSeriesId(metric, tags);
+
+    indexingStrategy.indexTimeseriesId(metric, tags, timeSeriesId);
+  }
+}
```
Comment (jzeeck) [inline cassandra/src/test/java/se/tre/freki/storage/cassandra/IndexingStrategyTest.java]: Should we have a test where this succeed as well?

File: cassandra/src/test/java/se/tre/freki/storage/cassandra/CassandraStoreTest.java [hunk 5]
```diff
@@ -66,13 +70,17 @@ public void constructorNullCluster() throws IOException {
...
```
Comment (jzeeck) [inline cassandra/src/test/java/se/tre/freki/storage/cassandra/CassandraStoreTest.java]: insertTagsStatement

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/CassandraStoreDescriptor.java [hunk 6]
```diff
@@ -74,7 +74,11 @@ public CassandraStore createStore(final Config config,
     final String keyspace = config.getString("freki.storage.cassandra.keyspace");
     final Session session = connectTo(cluster, keyspace);
     registerMetrics(cluster, metrics);
-    return new CassandraStore(cluster, session, Clock.systemDefaultZone());
+
+    final IndexStrategy onlineIndexStrategy = indexStrategyFor(session,
+        config.getBoolean("freki.storage.cassandra.online_index"));
+
```
Comment (jzeeck) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/CassandraStoreDescriptor.java]: change configurename?

File: cassandra/src/main/resources/reference.conf (hunk 3 above)
Comment (jzeeck) [inline cassandra/src/main/resources/reference.conf]: index_on_add_point

**Pull request merged**